server/src/main/java/org/elasticsearch/common/settings/KeyStoreLoader.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.settings;
+
+import org.elasticsearch.cli.ExitCodes;
+import org.elasticsearch.cli.Terminal;
+import org.elasticsearch.cli.UserException;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.env.Environment;
+
+import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+/**
+ * Implementation of {@link SecureSettingsLoader} for {@link KeyStoreWrapper}
+ */
+public class KeyStoreLoader implements SecureSettingsLoader {
+    private Path keystoreBackupPath = null;
+
+    @Override
+    public SecureSettings load(Environment environment, Terminal terminal, AutoConfigureFunction<SecureString, Environment> autoConfigure)
+        throws Exception {
+        // See if we have a keystore already present
+        KeyStoreWrapper secureSettings = KeyStoreWrapper.load(environment.configFile());
+        // If there's no keystore or the keystore has no password, set an empty password
+        var password = (secureSettings == null || secureSettings.hasPassword() == false)
+            ? new SecureString(new char[0])
+            : new SecureString(terminal.readSecret(KeyStoreWrapper.PROMPT));
+
+        environment = autoConfigure.apply(password);
+        secureSettings = KeyStoreWrapper.bootstrap(environment.configFile(), () -> password);
+        password.close();
+
+        return secureSettings;
+    }
+
+    @Override
+    public AutoConfigureResult autoConfigure(
```

Comment:
Why is anything from security auto configuration moved from AutoConfigureNode? I think my previous comment still applies, the keystore loader shouldn't need to know anything about security auto config.